new section: 3.1.3 Changing geometry, from Andre Sailer tutorial (#104)
HEP-FCC/fcc-tutorials: full-detector-simulations/FCCeeCLD/FCCeeCLD.md

To illustrate this workflow we use a simple but significat example: single taus generated with the particle gun.
We generate this with `k4run` and we save the result in `stdHEP` format for later processing in `DDSim`. 

## Changing geometry 

The description of the componentes of a particle detector can be extremely complex. `DD4hep` exploits the properties of `xml` file format to store the geometry and materials of each component of a given detector hierarchically. Thus the detector description is given in one or more `xml` files.

 In this section, the following topics are covered:
* how tow to simulate an existing detector model
* modify its XML files
* check for overlaps with Geant4
* get a better understand of what DD4hep does under the hood
* Write your own C++ detector constructor and run with it

This section is adapted from the [talk given by Andre Sailer](https://indico.cern.ch/event/1182767/contributions/5093486/attachments/2532164/4356996/221020_soft_tutorial_dd4hep.pdf) in the *FCC Software Tutorial 2022*.

### Setup

> The first step is to setup a CentOS 7 machine with CVMFS. It can be a local machine, remote machine (e.g. [lxplus.cern.ch](https://lxplusdoc.web.cern.ch/)), virtual machine (e.g. [CernVM](https://cernvm.cern.ch/appliance/)) or containerized system (e.g. [Docker](https://linux.web.cern.ch/dockerimages/), Singularity). 

Reviewer: The sourcing of the stack is no longer here

> The first step is to setup a CentOS 7 machine with CVMFS. It can be a local machine, remote machine (e.g. [lxplus.cern.ch](https://lxplusdoc.web.cern.ch/)), virtual machine (e.g. [CernVM](https://cernvm.cern.ch/appliance/)) or containerized system (e.g. [Docker](https://linux.web.cern.ch/dockerimages/), Singularity). 

Reviewer: ```suggestion
The first step is to set up a CentOS 7 machine with CVMFS. It can be a local machine, a remote machine (e.g. [lxplus.cern.ch](https://lxplusdoc.web.cern.ch/)), a virtual machine (e.g. [CernVM](https://cernvm.cern.ch/appliance/)) or a containerized system (e.g. [Docker](https://linux.web.cern.ch/dockerimages/) or Apptainer). 
```

The following step is to create a working directory,

```shell
mkdir mydd4heptutorial
cd mydd4heptutorial
cp -r $LCGEO/FCCee/compact/FCCee_o1_v05 .

> pip install --user uproot3

Reviewer: We could separate this pip install step

```

If everything goes fine, the contents of your working directory are like this:

```shell
ls -ltra
total 8
drwxr-xr-x. 28 sailer zf 4096 Oct 14 11:29 ..
drwxr-xr-x. 3 sailer zf 26 Oct 14 11:30 .
drwxr-xr-x. 2 sailer zf 4096 Oct 14 11:30 FCCee_o1_v05
```

> Another ingredient is a script to plot the key information after the simulations. [This python script](showPlots.py) plots the energy deposition and position along Z axis during the simulation. Histograms are printed in the terminal to avoid the use of the graphical interface. Conventional ROOT plot can be used instead with small modifications of the script.

Reviewer: ```suggestion
Another ingredient is a script to plot some information after the simulations to show that the geometry really did change. [This python script](showPlots.py) plots the per hit energy deposition and position along Z axis in the ECal endcap. The histograms are printed in the terminal to avoid the use of a graphical interface. If a graphical system is available, one can also visualize the geometry with `geoDisplay` or the Geant4 visualizations.
```

<details>

<summary><b>Click to show the code of the python script <code>showPlots.py</code> </b></summary>

```python
#!/bin/env python
import sys, ROOT, uproot3 as uproot
from ROOT import TFile, TTree
ROOT.gROOT.SetBatch(True)
if len(sys.argv) < 2:
  print("Please specify input file"); sys.exit(1)
inputFile = sys.argv[1]; print("Reading:", inputFile)
tfile = ROOT.TFile.Open(inputFile)
myTree = tfile.Get("events")
myTree.Draw("ECalEndcapCollection.position.z>>zHist(100, 2300, 2510)",
"ECalEndcapCollection.position.z > 0")
myTree.Draw("ECalEndcapCollection.energy>>eHist(30, 0, 0.002)")
zHist = tfile.Get("zHist")
eHist = tfile.Get("eHist")
oFile = ROOT.TFile.Open("temp.root", "RECREATE")
eHist.Write(); zHist.Write(); oFile.Close()
uFile = uproot.open("temp.root")
uFile["zHist"].show()
uFile["eHist"].show()
```

</details>

<br/>

> In this section, detector geometry will be changed in every step. Geometry can be visualized using different tools, as it is explained extensively in a following section. 

Reviewer: ```suggestion
In this section, the detector geometry will be changed in every step. The geometry can be visualized using different tools, as it is explained in the next section. 
```

> :::{admonition} Click to show some methods to plot the geometry.

Reviewer: ```suggestion
:::{admonition} Click to show some methods to display the geometry.
```

<details>

<summary><b>Click to show the code of the python script <code>dd4hep2root.py</code>. </b></summary>

```python
#!/usr/bin/env python3

import sys
import argparse

def main():
    parser = argparse.ArgumentParser(description='Convert detector')
    parser.add_argument('-c', '--compact', help='Compact file location(s)',
                        required=True, type=str, nargs='+')
    parser.add_argument('-o', '--out', help='Converted file path',
                        default='detector.root', type=str)
    args = vars(parser.parse_args())

    convert(args['compact'], args['out'])

def convert(compact_files, out_path):
    print('INFO: Converting following compact file(s):')
    for cfile in compact_files:
        print('      ' + cfile)

    import ROOT

    ROOT.gSystem.Load('libDDCore')
    description = ROOT.dd4hep.Detector.getInstance()
    for cfile in compact_files:
        description.fromXML(cfile)

    ROOT.gGeoManager.SetVisLevel(9)
    ROOT.gGeoManager.SetVisOption(0)
    ROOT.gGeoManager.Export(out_path)

if __name__ == '__main__':
    main()

```

</details>

<br/>

> 2. Using Qt interface of Geant4, accesible via `runType` argument of `ddsim`

Reviewer: ```suggestion
2. Using the `Qt` interface of Geant4, accessible via the `runType` argument of `ddsim`
```
As far as I remember, this doesn't work with the Key4hep stack at the moment. So explaining this as part of the tutorial is mighty confusing if someone where to actually try this.

<details>

<summary><b> Click to show how to display the detector geometry using Geant4. </b></summary>

```shell

> source /cvmfs/sft.cern.ch/lcg/views/dev4/latest/x86_64-centos7-gcc11-opt/setup.sh

Reviewer: Mixing different environments in one tutorial is very bad. If someone were to follow this step by step things wouldn't work.

One should move the displaying of geometry to a different tutorial and make clear that this has to be done in a clean terminal.

Author: I will move the visualization section inside the full detector simulation

> ddsim --compactFile description.xml --runType qt

Reviewer: The example description.xml file should be concrete and copy-pastable for the users.

```

> After building internally the geometry, a window will open. The menu on the left gives access to the builtin Geant4 macros. In order to show the geometry, we have to type some commands:

Reviewer: ```suggestion
After building the geometry, a window will open. The menu on the left gives access to the builtin Geant4 macros. In order to show the geometry, we have to type some commands:
```

```
# Use this open statement to create an OpenGL view:
/vis/open OGL 600x600-0+0

# Draw geometry:
/vis/drawVolume

# Specify view angle:
/vis/viewer/set/lightsVector -1 0 0
/vis/viewer/set/viewpointVector -1 0 0
```

</details>
:::

<br/>

### Simulate Muons in CLD detector

To simulate the interaction of 100 muons with the CLD detector, the following command is used inside the directory `mydd4heptutorial`. It takes some minutes to build the geometry and run the actual simulation. 

```shell
ddsim --compactFile FCCee_o1_v05/FCCee_o1_v05.xml \
      --enableGun \
      --gun.distribution uniform \
      --gun.energy "10*GeV" \
      --gun.particle mu- \
      --numberOfEvents 100 \
      --outputFile Step1_edm4hep.root
```

And now plot the distribution with `showPlots.py`

```shell
python showPlots.py Step1_edm4hep.root
```

### Removing an included XML file

> The detector geometry is complex, and takes a while to be built internally. To speed up the following steps, the Silicon Tracker can be removed. To do so, 

Reviewer: ```suggestion
The detector geometry is complex, and takes a while to be built. To speed up the following steps, the Silicon Tracker can be removed. To do so, 
```

It kind of makes sense to give the warning at the start and drop the tracker before any first simulation

1. Open `FCCee_o1_v05/FCCee_o1_v05.xml`

> 2. Find the following lines in the file, and replace/delete them

Reviewer: ```suggestion
2. Find the following lines in the file, and replace/delete them
```

replace them with what?

```xml
<include ref="InnerTracker_o2_v06_02.xml"/>
<include ref="OuterTracker_o2_v06_02.xml"/>
```

3. Comment out the plugins section `<plugins> ... </plugins>` in the same file. 

If now the simulation is run again as in the previous step (changing the output filename), 

```shell
ddsim --compactFile FCCee_o1_v05/FCCee_o1_v05.xml \
      --enableGun \
      --gun.distribution uniform \
      --gun.energy "10*GeV" \
      --gun.particle mu- \
      --numberOfEvents 100 \
      --outputFile Step2_edm4hep.root
```

The simulation runs 20 times faster and the distributions have not noticeably changed.

### Overlap checking

Whenever you change the geometry in a non-trivial way there are the possibilities of overlaps and the following things should be kept in mind

1. There are no trivial changes
2. See point 1
3. Run the overlap check

> Geant4 performs the overlap check of a given geometry. This action is available for `ddsim`. A a macro file with the Geant4 instructions has to be provided to `ddsim`. To do so, create a file named as *overlap.mac* and write these commands inside

Reviewer: ```suggestion
Geant4 can perform the overlap check of a given geometry. This action is available for `ddsim`. A macro file with the Geant4 instructions has to be provided to `ddsim`. To do so, create a file named as *overlap.mac* and write these commands inside
```
One can also give the commands manually, but if the output is too long it will be lost, so we use a macro file so that we can redirect the output and end ddsim gracefully at the end.

```
/geometry/test/run
exit
```

And then we run `ddsim` with this macro file, and dump the output to a text file for easy browsing:

```shell
ddsim --compactFile FCCee_o1_v05/FCCee_o1_v05.xml \
      --runType run \

>       --macroFile overlap.mac 2>&1 | tee overlapDump 

Reviewer: I don't like the `tee` here. If we just move the command to background we can already `less` the output file without losing everything we saw in the terminal before.

```

With the full detector model including the tracker this would take about 30 minutes. The output shows some exceptions:

* The exceptions about the VertexEndcapModule are due to a too small envelope, and could be
easily fixed

* The exceptions about the Boolean Volume are more vexing, let’s ignore those for now. The HOMAbsorber include can be dropped as well to silent these errors.

### Modifying the ECAL

In this section, it is shown how to change the number of layers and silicon thicknesses of the ECal Endcap. To do so,

1. Open the file `FCCee_o1_v05/ECalEndcap_o2_v01_03.xml`
2. Find this block:
```xml
<layer repeat="40" vis="ECalLayerVis">
<slice material = "TungstenDens24" thickness = "1.90*mm" vis="ECalAbsorberVis" radiator="yes"/>
<slice material = "G10" thickness = "0.15*mm" vis="InvisibleNoDaughters"/>
<slice material = "GroundOrHVMix" thickness = "0.10*mm" vis="ECalAbsorberVis"/>
<slice material = "Silicon" thickness = "0.50*mm" sensitive="yes" limits="cal_limits"

> vis="ECalSensitiveVis"/>

Reviewer: indentation at least to make it more obvious this is part of the `slice`

<slice material = "G10" thickness = "0.15*mm" vis="InvisibleNoDaughters"/>
<slice material = "GroundOrHVMix" thickness = "0.10*mm" vis="ECalAbsorberVis"/>
<slice material = "Silicon" thickness = "1.00*mm" sensitive="yes" limits="cal_limits"
vis="ECalSensitiveVis"/>
<slice material = "Air" thickness = "0.10*mm" vis="InvisibleNoDaughters"/>
<slice material = "siPCBMix" thickness = "1.30*mm" vis="ECalAbsorberVis"/>
<slice material = "Air" thickness = "0.25*mm" vis="InvisibleNoDaughters"/>
<slice material = "G10" thickness = "0.75*mm" vis="InvisibleNoDaughters"/>
</layer>
```

</details>

<br/>

Simulate again with the new ECal,

```shell
ddsim --compactFile FCCee_o1_v05/FCCee_o1_v05.xml \
      --enableGun \
      --gun.distribution uniform \
      --gun.energy "10*GeV" \
      --gun.particle mu- \
      --numberOfEvents 100 \
      --outputFile Step3_edm4hep.root
```

Compare the histograms produced from `Step2_edm4hep.root` and `Step3_edm4hep.root`.

### Geometry driver modifications

The type attribute of the detector tag tells DD4hep which detector constructor to load

```xml
<detector name="ECalEndcap"

> type="GenericCalEndcap_o1_v01"

Reviewer: indendations?

                offset_r="ECalBarrel_inner_radius"
                grid_size_r="1*cm" />
            <id>system:5,side:2,module:8,stave:4,layer:9,submodule:4,r:32:10,eta:-11,phi:-11</id>
        </readout>
    </readouts>

    <display>
        <vis name="MyVis" alpha="0.1"
            r="0.1" g=".5" b=".5"
            showDaughters="true"
            visible="false" />
    </display>

    <detectors>
        <detector name="MyDetectorName"
            type="MyFirstDetector"
            id="1234"
            readout="MyReadout"
            vis="MyVis">
            <dimensions
                zmax="ECalBarrel_half_length"
                rmin="ECalBarrel_inner_radius"
                rmax="ECalBarrel_outer_radius" />
        </detector>
    </detectors>
    
</lccdd>
```

</details>

<br/>

Now, modify the `FCCee_o1_v05/FCCee_o1_v05.xml` and replace the include for the `ECalBarrel`
with `mydetector.xml`. After every change in the geometry, the overlap check must be run inside `FCCee_o1_v05` directory 

```shell
ddsim --compactFile FCCee_o1_v05/FCCee_o1_v05.xml \
      --runType run \
      --macroFile overlap.mac 2>&1 | tee overlapDump2 
```

> If there are no conflict, we can continue. Now, we add some layers to the detector. First, we add a `layer` component inside the `<detector> ... </detector>` section of `FCCee_o1_v05/mydetector.xml` detector description file.

Reviewer: ```suggestion
If there are no exceptions, we can continue. Now, we add some layers to the detector. First, we add a `layer` component inside the `<detector> ... </detector>` section of `FCCee_o1_v05/mydetector.xml` detector description file.
```
